spec/requests/orders_spec.rb

```diff
@@ -0,0 +1,143 @@
+  let!(:product) { create(:product, status: "enabled") }
+
+  before do
+    sign_in user
+  end
+
+  describe "GET #index" do
+    let!(:order) { create :order, user: user, status: status }
+    
+    context "when get all orders success" do
+      let(:status) { :wait_confirm }
+
+      before { get :index }
+
+      it do
+        expect(response).to be_successful
+        expect(response).to render_template "index"
+        expect(assigns[:orders]).to eq [order]
+      end
+    end
+
+    context "when get orders by status" do
+      let(:status) { :delivering }
+
+      before { get :index, params: { status: status} }
+
+      it do
+        expect(response).to be_successful
+        expect(response).to render_template "index"
+        expect(assigns[:orders]).to eq [order]
+        expect(assigns[:orders].first.status).to eq status.to_s
+      end
+    end
+  end
+
+  describe "POST #create" do
+    subject { post :create, params: { order: order_params } }
+
+    let(:cart_items) { { product.id => 1 }.to_json }
+    let(:order_params) { { address: "123 Main St", phone: "123-456-7890" } }
+
+    before { request.cookies[:cart_items] = cart_items }
+
+    after { clear_enqueued_jobs }
+
+    describe "callbacks" do
+      describe "#cart_items" do
+        it do
+          expect { subject }.to change { assigns(:cart_items) }
+                    .from(assigns(:cart_items)).to(JSON.parse(cart_items))
+        end
+      end
+    end
+
+    context "when valid product and order params" do
+      let(:order_mailer_jobs) do
+        ActiveJob::Base.queue_adapter.enqueued_jobs.select do |job|
+          job[:args][0] == "OrderMailer" && job[:args][1] == "place_order"
+        end
+      end
+
+      it do
+        expect { subject }.to change(Order, :count).by(1)
+        expect(response).to redirect_to orders_path
+        expect(response.cookies).to include("cart_items" => nil)
+        expect(flash[:success]).to eq "Order successfully placed! Please check your email to track the notification"
+        expect(order_mailer_jobs.size).to eq 1
+      end
+    end
+
+    context "when invalid product" do
+      let(:cart_items) { { "4" => 1 }.to_json }
+      let(:order_params) { { address: "123 Main St", phone: "123-456-7890" } }
+
+      it do
+        expect { subject }.not_to change(Order, :count)
+        expect(response).to redirect_to cart_path
+        expect(flash[:error]).to eq "Product not found."
+        expect(response.cookies).to include("cart_items" => "{}")
+      end
+    end
+
+    context "when invalid order params" do
+      let(:order_params) { { address: "123 Main St" } }
+
+      it do
+        expect { subject }.not_to change(Order, :count)
+        expect(flash[:error]).to eq "Order failed! Please fill out your address and phone"
+        expect(response).to redirect_to cart_path
+      end
+    end
+  end
+
+  describe "PATCH #update" do
+    subject { patch :update, params: { 
+                id: order,
```

Comment:
ntn mới đúng nhi
thêm 1 `let(:order_id)` để resue ở phía dưới nha k nên viết lại subject
```suggestion
                id: order.id,
```

Reply:
@taylp-2865 dạ okie anh để em fix lại